django/db/models/fields/files.py

```diff
@@ -91,7 +91,10 @@ def open(self, mode="rb"):
     def save(self, name, content, save=True):
         name = self.field.generate_filename(self.instance, name)
         self.name = self.storage.save(name, content, max_length=self.field.max_length)
-        setattr(self.instance, self.field.attname, self.name)
+        setattr(self.instance, self.field.attname, content)
+        # Update the name in case generate_filename() or storage.save() changed
+        # it, but bypass the descriptor to avoid re-reading the file.
+        self.instance.__dict__[self.field.attname] = self.name
```

Comment:
Right now we have updated `save` in `FieldFile` due to an issue in `ImageField` due to the complexity of `ImageFileDescriptor` and `ImageField`. Then we documented the potential impact to `ImageField` but actually all FieldFields might be affected. So I think we should isolate the solution to the problem being solved.

Easy option is to overwrite the `save`  of `ImageFieldFile`.
Later I would love us to simplify `update_dimension_fields` which I think is a bit of a mess

Reply:
Presumably the concern is that someone out there has subclassed FileField and relies on the descriptor getting called in this specific case?

Reply:
Let's supposed you have a Model with a FileField, not an image field

```python3
class Model(models.Model):
    file = models.FileField(...)
```

And you do the following:

```python
instance = Model()
instance.file.save("myfile", ContentFile(b"foo"))
instance.file.read()
```

BEFORE my change this would result in the file being read off the storage backend. AFTER this would result in directly reading the value from `ContentFile`.

So you're right that there's a change in behavior for non-image fields, but I'm wondering if that's the more "correct" or "desirable" behavior in the general case.

Reply:
I just deleted a comment because I'm not sure it's right, and need to double check some things.

I'm wondering what happens if you call `read()` on a filefield after saving it: does it hit the storage backend again or not?

Reply:
OK, so with this patch, reading a file immediately after saving it will involve hitting the storage backend, even if the file _could_ be cached in memory

Example code:


```python
class Model(models.Model):
    file = models.FileField(...)

instance = Model()
instance.file = ContentFile(b"foo", name="bar.txt")
instance.file.read() == b"foo"  # No read
instance.save()
# instance.__dict__["file"] is now an instance of str() and the underlying ContentFile is dicsarded
instance.file.read() == b"foo"  # Will always result in a read to the storage backend
```

Reply:
If we guard the code like this


```python3
  setattr(self.instance, self.field.attname, content)
  # Update the name in case generate_filename() or storage.save() changed
  # it, but bypass the descriptor to avoid re-reading the file.
  # Only update in the case that it's strictly necessary
  if getattr(self.instance, self.field.attname).name != self.name:
      self.instance.__dict__[self.field.attname] = self.name
 ```
 
 Now, the flow control is more complicated, and there's two different scenarios
 
 ```python
instance = Model()
instance.file = ContentFile(b"foo", name="bar.txt")
instance.file.read() == b"foo"  # No hit to the storage backend
instance.save()
# If the storage backend is "cool" with "bar.txt" as a filename then
# instance.__dict__["file"] is not an instance of str(), but rather a file-like object
instance.file.read() == b"foo"  # Won't necessarily hit the storage backend
```

However, if the storage backend changes the filename for any reason (either the file already exists, it normalized whitespace, literally any reason), then it will result in the final read() hitting the storage backend.

Reply:
I've got something like this

```python3
    def _set_instance_attribute(self, name, content):
        # Existence of this method is primarily motivated by custom logic
        # in ImageFieldFile._set_instance_attribute
        setattr(self.instance, self.field.attname, name)

    def save(self, name, content, save=True):
        name = self.field.generate_filename(self.instance, name)
        self.name = self.storage.save(name, content, max_length=self.field.max_length)
        self._set_instance_attribute(self.name, content)
        self._committed = True

        # Save the object because it has changed, unless save is False
        if save:
            self.instance.save()
  ```
  
  And then in `ImageFieldFile`
  
  
  ```python3
      def _set_instance_attribute(self, name, content):
        setattr(self.instance, self.field.attname, content)
        # Update the name in case generate_filename() or storage.save() changed
        # it, but bypass the descriptor to avoid re-reading the file.
        self.instance.__dict__[self.field.attname] = self.name
   ```
   
   That way I don't have to just wholesale copy/paste the save method when overwriting it

Reply:
Nice idea 💡

Reply:
@sarahboyce Updated and force-pushed